File: ios/RoadRank/Services/HapticManager.swift
import UIKit

// MARK: - Haptic Manager
final class HapticManager {
    static let shared = HapticManager()

    private let impactLight = UIImpactFeedbackGenerator(style: .light)
    private let impactMedium = UIImpactFeedbackGenerator(style: .medium)
    private let impactHeavy = UIImpactFeedbackGenerator(style: .heavy)
    private let impactSoft = UIImpactFeedbackGenerator(style: .soft)
    private let impactRigid = UIImpactFeedbackGenerator(style: .rigid)
    private let selectionGenerator = UISelectionFeedbackGenerator()
    private let notificationGenerator = UINotificationFeedbackGenerator()

    private init() {
        // Prepare generators for faster response
        prepareGenerators()
    }

    private func prepareGenerators() {
        impactLight.prepare()
        impactMedium.prepare()
        impactHeavy.prepare()
        impactSoft.prepare()
        impactRigid.prepare()
        selectionGenerator.prepare()
        notificationGenerator.prepare()
    }

    // MARK: - Impact Feedback

    func impact(_ style: UIImpactFeedbackGenerator.FeedbackStyle) {
        switch style {
        case .light:
            impactLight.impactOccurred()
        case .medium:
            impactMedium.impactOccurred()
        case .heavy:
            impactHeavy.impactOccurred()
        case .soft:
            impactSoft.impactOccurred()
        case .rigid:
            impactRigid.impactOccurred()
        @unknown default:
            impactMedium.impactOccurred()
        }
        prepareGenerators()
    }

    func impact(_ style: UIImpactFeedbackGenerator.FeedbackStyle, intensity: CGFloat) {
        switch style {
        case .light:
            impactLight.impactOccurred(intensity: intensity)
        case .medium:
            impactMedium.impactOccurred(intensity: intensity)
        case .heavy:
            impactHeavy.impactOccurred(intensity: intensity)
        case .soft:
            impactSoft.impactOccurred(intensity: intensity)
        case .rigid:
            impactRigid.impactOccurred(intensity: intensity)
        @unknown default:
            impactMedium.impactOccurred(intensity: intensity)
        }
        prepareGenerators()
    }

    // MARK: - Selection Feedback

    func selection() {
        selectionGenerator.selectionChanged()
        selectionGenerator.prepare()
    }

    // MARK: - Notification Feedback

    func notification(_ type: UINotificationFeedbackGenerator.FeedbackType) {
        notificationGenerator.notificationOccurred(type)
        notificationGenerator.prepare()
    }

    // MARK: - Custom Patterns

    func success() {
        notification(.success)
    }

    func warning() {
        notification(.warning)
    }

    func error() {
        notification(.error)
    }

    func buttonTap() {
        impact(.light)
    }

    func toggle() {
        impact(.medium)
    }

    func sliderChange() {
        selection()
    }

    func longPress() {
        impact(.heavy)
    }

    func draw() {
        impact(.soft, intensity: 0.5)
    }
}

File: android/app/src/main/java/com/roadrank/app/services/HapticManager.kt
package com.roadrank.app.services

import android.content.Context
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import androidx.core.content.getSystemService

/**
 * Haptic Manager for haptic feedback - matching iOS
 */
object HapticManager {
    private var vibrator: Vibrator? = null

    fun init(context: Context) {
        vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val vibratorManager = context.getSystemService<VibratorManager>()
            vibratorManager?.defaultVibrator
        } else {
            @Suppress("DEPRECATION")
            context.getSystemService(Context.VIBRATOR_SERVICE) as? Vibrator
        }
    }

    enum class ImpactStyle(val duration: Long, val amplitude: Int) {
        LIGHT(20L, 50),
        MEDIUM(30L, 100),
        HEAVY(40L, 180),
        SOFT(15L, 30),
        RIGID(25L, 255)
    }

    enum class NotificationType(val pattern: LongArray, val amplitudes: IntArray) {
        SUCCESS(longArrayOf(0, 50, 50, 50), intArrayOf(0, 100, 0, 150)),
        WARNING(longArrayOf(0, 50, 100, 50), intArrayOf(0, 150, 0, 100)),
        ERROR(longArrayOf(0, 100, 50, 100), intArrayOf(0, 200, 0, 200))
    }

    fun impact(style: ImpactStyle) {
        vibrator?.let { vib ->
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                vib.vibrate(
                    VibrationEffect.createOneShot(style.duration, style.amplitude)
                )
            } else {
                @Suppress("DEPRECATION")
                vib.vibrate(style.duration)
            }
        }
    }

    fun selection() {
        impact(ImpactStyle.LIGHT)
    }

    fun notification(type: NotificationType) {
        vibrator?.let { vib ->
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                vib.vibrate(
                    VibrationEffect.createWaveform(type.pattern, type.amplitudes, -1)
                )
            } else {
                @Suppress("DEPRECATION")
                vib.vibrate(type.pattern, -1)
            }
        }
    }

    // Convenience methods matching iOS
    fun success() = notification(NotificationType.SUCCESS)
    fun warning() = notification(NotificationType.WARNING)
    fun error() = notification(NotificationType.ERROR)
    fun buttonTap() = impact(ImpactStyle.LIGHT)
    fun toggle() = impact(ImpactStyle.MEDIUM)
    fun sliderChange() = selection()
    fun longPress() = impact(ImpactStyle.HEAVY)
    fun draw() = impact(ImpactStyle.SOFT)
}
